### PR TITLE
param_shmem sync with param

### PR DESCRIPTION
**Pull request by dagar**
The next step is merging this.

### CODE REVIEW

**Comment by dagar** [comment]
@bkueng this is low priority, but I was thinking we should merge param_shmem into the regular param. This PR gets them close to in sync again.

File: src/modules/systemlib/param/param_shmem.c [hunk 1]
```diff
@@ -162,32 +165,83 @@ get_param_info_count(void)
...
 	// TODO: this doesn't seem to work on Snappy
 	//do {} while (px4_sem_wait(&param_sem) != 0);
 }
 
 /** unlock the parameter store */
 static void
-param_unlock(void)
+param_unlock_reader(void)
+{
+	// TODO: this doesn't seem to work on Snappy
```
Comment (bkueng) [inline src/modules/systemlib/param/param_shmem.c]: we will need to fix this

File: src/modules/systemlib/param/param_shmem.c [hunk 2]
```diff
@@ -1264,34 +1395,35 @@ static int
...
-		goto out;
+		PX4_ERR("decoder init failed");
+		return PX4_ERROR;
 	}
 
 	state.mark_saved = mark_saved;
 
 	do {
 		result = bson_decoder_next(&decoder);
+		usleep(1);
```
Comment (bkueng) [inline src/modules/systemlib/param/param_shmem.c]: I think we can remove this.

**Comment by LorenzMeier** [comment]
@DanielePettenuzzo Could you please test this with Nico? This might very well resolve the boot issue you had.

**Comment by DanielePettenuzzo** [comment]
@dagar @LorenzMeier we have this boot issue also with this pr. 
The last working firmware commit  is: 76bf9c6
Now we have an issue for this: https://github.com/PX4/Firmware/issues/9444

**Pull request closed without merging**